File: samples/dnn/face_recognizer.cpp
#include <iostream>
#include <vector>

#include <opencv2/dnn.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>

constexpr double COSINE_THRESHOLD = 0.363;
constexpr double L2NORM_THRESHOLD = 1.128;

void visualize(
    cv::Mat& image1,
    const cv::Rect box1,
    cv::Mat& image2,
    const cv::Rect box2,
    const double score,
    const cv::dnn::FaceRecognitionModel::DisType distance_type = cv::dnn::FaceRecognitionModel::DisType::FR_COSINE,
    const int thickness = 2
)
{
    // Green is same identity, Red is different identities.
    bool same_identity = false;
    if (distance_type == cv::dnn::FaceRecognitionModel::DisType::FR_COSINE)
    {
        same_identity = (score >= L2NORM_THRESHOLD);
    }
    else{
        same_identity = (score <= L2NORM_THRESHOLD);
    }
    color = same_identity ? cv::Scalar(0, 255, 0) : cv::Scalar(255, 0, 0);

    // Draw Bounding Box
    cv::rectangle(image1, box1, color, thickness);
    cv::rectangle(image2, box2, color, thickness);
}

int main(int argc, char** argv)
{
    cv::CommandLineParser parser(argc, argv,
        "{help h                    |                                     | Print this message                                                                                                                                       }"
        "{image1 i1                 |                                     | Path to the input image1. When image1 and image2 parameters given then the program try to find a face on both images and runs face recognition algorithm.}"
        "{image2 i2                 |                                     | Path to the input image2. When image1 and image2 parameters given then the program try to find a face on both images and runs face recognition algorithm.}"
        "{scale sc                  | 1.0                                 | Scale factor used to resize input video frames (0.0-1.0]                                                                                                 }"
        "{face_detection_model fd   | face_detection_yunet_2021dec.onnx   | Path to the model. Download yunet.onnx in https://github.com/opencv/opencv_zoo/tree/master/models/face_detection_yunet                                   }"
        "{face_recognition_model fr | face_recognition_sface_2021dec.onnx | Path to the face recognition model. Download the model at https://github.com/opencv/opencv_zoo/tree/master/models/face_recognition_sface                 }"
        "{score_threshold st        | 0.9                                 | Filter out faces of score < score_threshold                                                                                                              }"
        "{nms_threshold nt          | 0.3                                 | Suppress bounding boxes of iou >= nms_threshold                                                                                                          }"
    );
    if (argc == 1 || parser.has("help"))
    {
        parser.printMessage();
        std::cout << "e.g. face_recognizer.exe --image1=./image1.jpg --image2=./image2.jpg --face_detection_model=./face_detection_yunet_2021dec.onnx -face_recognition_model=./face_recognition_sface_2021dec.onnx" << std::endl;
        return 0;
    }
    if (!parser.has("image1") || !parser.has("image2"))
    {
        std::cerr << "can't specify image paths" << std::endl;
        return -1;
    }

    const float scale = parser.get<float>("scale");
    const float score_threshold = parser.get<float>("score_threshold");
    const float nms_threshold = parser.get<float>("nms_threshold");

    const cv::String face_detection_model_path = parser.get<cv::String>("face_detection_model");
    cv::dnn::FaceDetectionModel_YN face_detector = cv::dnn::FaceDetectionModel_YN(face_detection_model_path);

    //! [Create_FaceRecognitionModel_SF]
    // Create FaceRecognitionModel_SF
    const cv::String face_recognition_model_path = parser.get<cv::String>("face_recognition_model");
    cv::dnn::FaceRecognitionModel_SF face_recognizer = cv::dnn::FaceRecognitionModel_SF(face_recognition_model_path);
    //! [Create_FaceRecognitionModel_SF]

    const cv::String image1_path = parser.get<cv::String>("image1");
    cv::Mat image1 = cv::imread(cv::samples::findFile(image1_path));
    if (image1.empty())
    {
        std::cerr << "can't read image1: " << image1_path << std::endl;
        return -1;
    }

    cv::resize(image1, image1, cv::Size(), scale, scale);

    std::vector<float> confidences1;
    std::vector<cv::Rect> boxes1;
    face_detector.detect(image1, confidences1, boxes1, score_threshold, nms_threshold);

    std::vector<std::vector<cv::Point>> landmarks1;
    face_detector.getLandmarks( landmarks1 );

    if (landmarks1.size() == 0)
    {
        std::cerr << "can't detect face from image1" << std::endl;
        return -1;
    }

    const cv::String image2_path = parser.get<cv::String>("image2");
    cv::Mat image2 = cv::imread(cv::samples::findFile(image2_path));
    if (image2.empty())
    {
        std::cerr << "can't read image2: " << image2_path << std::endl;
        return -1;
    }

    cv::resize(image2, image2, cv::Size(), scale, scale);

    std::vector<float> confidences2;
    std::vector<cv::Rect> boxes2;
    face_detector.detect(image2, confidences2, boxes2, score_threshold, nms_threshold);

    std::vector<std::vector<cv::Point>> landmarks2;
    face_detector.getLandmarks( landmarks2 );

    if (landmarks2.size() == 0)
    {
        std::cerr << "can't detect face from image2" << std::endl;
        return -1;
    }

    //! [Align_Crop]
    // Aligning and Cropping facial image through the first face of faces detected.
    // In the case of SFace, It use five landmarks that lined up in a specific order.
    // (Right-Eye, Left-Eye, Nose, Right-Mouth Corner, Right-Mouth Corner)
    cv::Mat aligned_face1, aligned_face2;
    face_recognizer.alignCrop(image1, aligned_face1, landmarks1[0]);
    face_recognizer.alignCrop(image2, aligned_face2, landmarks2[0]);
    //! [Align_Crop]

    //! [Extract_Feature]
    // Run feature extraction with given aligned_face.
    cv::Mat face_feature1, face_feature2;
    face_recognizer.feature(aligned_face1, face_feature1);
    face_recognizer.feature(aligned_face2, face_feature2);
    //! [Extract_Feature]

    //! [Match_Features]
    // Match two features using each distance types.
    // * DisType::FR_COSINE : Cosine similarity. Higher value means higher similarity. (max 1.0)
    // * DisType::FR_NORM_L2 : L2-Norm distance. Lower value means higher similarity. (min 0.0)
    const double cos_score = face_recognizer.match(face_feature1, face_feature2, cv::dnn::FaceRecognitionModel::DisType::FR_COSINE);
    const double l2norm_score = face_recognizer.match(face_feature1, face_feature2, cv::dnn::FaceRecognitionModel::DisType::FR_NORM_L2);
    //! [Match_Features]

    //! [Check_Identity]
    // Check identity using Cosine similarity.
    if (cos_score >= COSINE_THRESHOLD)
    {
        std::cout << "They have the same identity;";
    }
    else
    {
        std::cout << "They have different identities;";
    }
    std::cout << " Cosine Similarity: " << cos_score << ", threshold: " << COSINE_THRESHOLD << ". (higher value means higher similarity, max 1.0)" << std::endl;

    // Check identity using L2-Norm distance.
    if (l2norm_score <= L2NORM_THRESHOLD)
    {
        std::cout << "They have the same identity;";
    }
    else
    {
        std::cout << "They have different identities.";
    }
    std::cout << " L2-Norm Distance: " << l2norm_score << ", threshold: " << L2NORM_THRESHOLD << ". (lower value means higher similarity, min 0.0)" << std::endl;
    //! [Check_Identity]

    visualize(image1, boxes1[0], image2, boxes2[0], cos_score);

    cv::imshow("face recognizer (image1)", image1);
    cv::imshow("face recognizer (image2)", image2);
    cv::waitKey(0);

    return 0;
}

File: samples/dnn/face_recognizer.py
import argparse

import numpy as np
import cv2 as cv

COSINE_THRESHOLD = 0.363
L2NORM_THRESHOLD = 1.128

# e.g. std::cout << "e.g. python3 face_recognizer.py --image1 ./image1.jpg --image2 ./image2.jpg --face_detection_model ./face_detection_yunet_2021dec.onnx -face_recognition_model ./face_recognition_sface_2021dec.onnx" << std::endl;
parser = argparse.ArgumentParser()
parser.add_argument('--image1', '-i1', type=str, help='Path to the input image1. When image1 and image2 parameters given then the program try to find a face on both images and runs face recognition algorithm.')
parser.add_argument('--image2', '-i2', type=str, help='Path to the input image2. When image1 and image2 parameters given then the program try to find a face on both images and runs face recognition algorithm.')
parser.add_argument('--scale', '-sc', type=float, default=1.0, help='Scale factor used to resize input video frames.')
parser.add_argument('--face_detection_model', '-fd', type=str, default='face_detection_yunet_2021dec.onnx', help='Path to the face detection model. Download the model at https://github.com/opencv/opencv_zoo/tree/master/models/face_detection_yunet')
parser.add_argument('--face_recognition_model', '-fr', type=str, default='face_recognition_sface_2021dec.onnx', help='Path to the face recognition model. Download the model at https://github.com/opencv/opencv_zoo/tree/master/models/face_recognition_sface')
parser.add_argument('--score_threshold', type=float, default=0.9, help='Filtering out faces of score < score_threshold.')
parser.add_argument('--nms_threshold', type=float, default=0.3, help='Suppress bounding boxes of iou >= nms_threshold.')
args = parser.parse_args()

def visualize(image1, box1, image2, box2, score, distance_type=cv.dnn.FaceRecognitionModel_FR_COSINE, thickness=2):
    # Green is same identity, Red is different identities.
    if distance_type == 0:
        same_identity = (score >= COSINE_THRESHOLD)
    else:
        same_identity = (score <= L2NORM_THRESHOLD)
    color = (0, 255, 0) if same_identity else (255, 0, 0)

    # Draw Bounding Box
    cv.rectangle(image1, box1, color, thickness)
    cv.rectangle(image2, box2, color, thickness)

if __name__ == '__main__':
    if (args.image1 is None) or (args.image2 is None):
        raise Exception("can't specify image paths")

    scale = args.scale
    score_threshold = args.score_threshold
    nms_threshold = args.nms_threshold

    face_detection_model_path = args.face_detection_model
    print(face_detection_model_path)
    face_detector = cv.dnn_FaceDetectionModel_YN(face_detection_model_path)

    ## [Create_FaceRecognitionModel_SF]
    # Create FaceRecognitionModel_SF
    face_recognition_model_path = args.face_recognition_model
    face_recognizer = cv.dnn_FaceRecognitionModel_SF(face_recognition_model_path)
    ## [Create_FaceRecognitionModel_SF]

    image1_path = args.image1
    image1 = cv.imread(cv.samples.findFile(image1_path))
    if image1 is None:
        raise Exception("can't read image1: {}".format(image1_path))

    image1 = cv.resize(image1, dsize=None, fx=scale, fy=scale)

    _, boxes1 = face_detector.detect(image1, score_threshold, nms_threshold)
    landmarks1 = face_detector.getLandmarks()
    if len(landmarks1) == 0:
        raise Exception("can't detect face from image1")

    image2_path = args.image2
    image2 = cv.imread(cv.samples.findFile(image2_path))
    if image2 is None:
        raise Exception("can't read image2: {}".format(image2_path))

    image2= cv.resize(image2, dsize=None, fx=scale, fy=scale)

    _, boxes2 = face_detector.detect(image2, score_threshold, nms_threshold)
    landmarks2 = face_detector.getLandmarks()
    if len(landmarks2) == 0:
        raise Exception("can't detect face from image2")

    ## [Align_Crop]
    # Aligning and Cropping facial image through the first face of faces detected.
    # In the case of SFace, It use five landmarks that lined up in a specific order.
    # (Right-Eye, Left-Eye, Nose, Right-Mouth Corner, Right-Mouth Corner)
    aligned_face1 = face_recognizer.alignCrop(image1, landmarks1[0])
    aligned_face2 = face_recognizer.alignCrop(image2, landmarks2[0])
    ## [Align_Crop]

    ## [Extract_Feature]
    # Run feature extraction with given aligned_face.
    face_feature1 = face_recognizer.feature(aligned_face1)
    face_feature2 = face_recognizer.feature(aligned_face2)
    ## [Extract_Feature]

    ## [Match_Features]
    # Match two features using each distance types.
    # * DisType::FR_COSINE : Cosine similarity. Higher value means higher similarity. (max 1.0)
    # * DisType::FR_NORM_L2 : L2-Norm distance. Lower value means higher similarity. (min 0.0)
    cos_score = face_recognizer.match(face_feature1, face_feature2, cv.dnn.FaceRecognitionModel_FR_COSINE)
    l2norm_score = face_recognizer.match(face_feature1, face_feature2, cv.dnn.FaceRecognitionModel_NORM_L2)
    ## [Match_Features]

    ## [Check_Identity]
    # Check identity using Cosine similarity.
    if (cos_score >= COSINE_THRESHOLD):
        print("They have the same identity;", end="")
    else:
        print("They have different identities;", end="")
    print(" Cosine Similarity: {}, threshold: {}. (higher value means higher similarity, max 1.0)".format(cos_score, COSINE_THRESHOLD));

    # Check identity using L2-Norm distance.
    if (l2norm_score <= L2NORM_THRESHOLD):
        print("They have the same identity;", end="")
    else:
        print("They have different identities;", end="")
    print(" L2-Norm Distance: {}, threshold: {}. (lower value means higher similarity, min 0.0)".format(l2norm_score, L2NORM_THRESHOLD))
    ## [Check_Identity]

    visualize(image1, boxes1[0], image2, boxes2[0], cos_score)

    cv.imshow("face recognizer (image1)", image1)
    cv.imshow("face recognizer (image2)", image2)
    cv.waitKey(0);
